### ISSUE TITLE
[bug] User friendly name isn't specific to the module that imports it in new spreadsheet generator

### ISSUE THREAD

**Issue by mshadbolt** [ebi-ait/hca-ebi-dev-team#321]
When the latest spreadsheet generator imports a module, the user friendly name is not being made specific to the module that it is related to.

For example if we create a sheet with the `organism_age`, `organism_age_unit`, `gestational_age`, `gestational_age_unit`, the top row user friendly names for these would be

| AGE | TIME UNIT | GESTATIONAL AGE | TIME UNIT |
| ---- | ----------- | ------------------- | ----------- |

Whereas each `TIME UNIT` field should be changed to `AGE UNIT` and `GESTATIONAL AGE UNIT`.

This most likely needs to be changed in the code that gets the user friendly name within the `ingest-client` module, `spreadsheet_builder.py`

**Comment by ami-day** [comment]
**Update:**

I have found the source of the problem but I am not yet sure how best to fix it. I am going to use the Donor organism species field as an example to illustrate.

When the "spreadsheet_generator.tab_for_type(type_spec)" function is run with the 'donor_organism" TypeSpec as input, a ParsedTab object is generated. This function is in the "spreadsheet_generator.py" script in the ingest-broker repo.

In the ParsedTab object, there are columns objects. In the columns object for the 'donor_organism' ParsedTab object, you can find the following:

path = donor_organism.genus_species.text
name = “Species”

Also, in the metadata_properties object in the schema_template object, there is a donor_organism object with the following:

genus_species.text.user_friendly = “Species"

This is why "Species" is being used as the user-friendly column name, rather than "Genus species".

However, in the donor_organism schema here: https://schema.humancellatlas.org/type/biomaterial/15.5.0/donor_organism,
the user friendly name that is specified is "Genus species". For example the path would be: donor_organsim.genus_species.user_friendly = “Genus species"

@mshadbolt @rdgoite others on devs it would be great to get some help on fixing this, if/when you have time.

**Comment by mshadbolt** [comment]
So when the tab is parsed, it gets the user-friendly name directly from the schema?

I am not really sure how, but in the old spreadsheet generator it was able to customise the user friendly names. And it seems that it works properly for some fields, i.e. the biomaterial core fields have user friendly names that relate to the type of biomaterial that imports it. Is there a difference in the way that the user friendly names are generated for biomaterial core fields vs ontology module fields? 

I am not that concerned about the 'species' one, more the ontology modules that aren't being typed, e.g. `TIME UNIT` `LENGTH UNIT` `MASS UNIT`

**Comment by mshadbolt** [comment]
so I think I have tracked the problem down further. When the schemaTemplate object is created, the json schemas themselves are parsed one way whereas when they are put into the metadata_properties dictionary they are parsed another way, for example:

If you look directly at the schema objects within the SchemaTemplate object the correct user friendly name is returned: 
```
test_template.json_schemas[24]['properties']['organism_age_unit']['user_friendly']
'Age unit'
```

whereas using the method 
```
test_template.lookup_property_attributes_in_metadata("donor_organism.organism_age_unit.text.user_friendly")
'Time unit'
```

**Comment by mshadbolt** [comment]
I have an initial attempt at fixing on this branch https://github.com/ebi-ait/ingest-client/blob/ms-fix-typed-ontology-uf/ingest/template/spreadsheet_builder.py

Seems to work locally but I want to add more tests, the current unit tests for the spreadsheet builder code are almost not existent

**Comment by mshadbolt** [comment]
hmm my attempt at simpler code breaks everything. Looks like I have to look into the objects in the descriptor code to ensure we get the user friendly name from the schema that imports a module rather than the module itself.

**Comment by mshadbolt** [comment]
rolando pushed a change to ingest-client so that we can access the userfriendly names in the source schema, I then made some updates to the 'get_user_friendly_name' method and it is working for some of the examples mentioned above but there are still some instances when it doesn't. e.g.
"cell_suspension.timecourse.unit.text": returns "Time unit" instead of "Timecourse unit" 
"specimen_from_organism.preservation_storage.storage_time_unit.text": returns "Time unit" instead of "Storage time unit"
"cell_suspension.cell_morphology.cell_size_unit.text": returns "Length unit" instead of "Cell size unit"
"donor_organism.genus_species.text": returns "Species" instead of "Genus species"

the branch we are working on is https://github.com/ebi-ait/ingest-client/tree/bug/user-friendly-names

**Comment by mshadbolt** [comment]
In the latest version of the branch above I have fixed the issue with the `(Required)` showing up on non-required fields. I traced this to the fact it was looking up the value for required in the ontology module (in ontology modules the text field is always required) and adjusted the 'get_value_from_column' function to behave differently when `.text` is in the column_name

I haven't been able to figure out why sometimes when ontologies are merged, the "user_friendly" is kept in the properties and sometimes it isn't. Since the change yesterday we can now lookup `donor_organism.organism_age_unit.user_friendly` and get the correct user_friendly name returned, but when you try to look up `donor_organism.genus_species.user_friendly` this field does not exist, even though it is present in the schema itself.

**Comment by mshadbolt** [comment]
tldr; Latest changes to the branch incorporate band-aid fixes for the issues above but don't address the root cause.

Basically anytime a module imports another module, the user friendly name is getting populated from the last imported module in the chain. I think this is due to the way the `jsonref` package follows the references in the schema, it follows the ref but doesn't retain any other information about that field aside from the literal `$ref` field.

The merging that we implemented works when it is a 2 level relationship, but doesn't work when it is 3 level.

I have basically added another exception into the `get_user_friendly_name` method but this method is getting pretty messy and it would be better in the long term to ensure that we are parsing the schema correctly to begin with and not losing information in the process.

The same issue affects the descriptive and example text, for example:
we have dedicated text for the 'method' ontology module when it is imported by the dissociation protocol schema:
> "The method by which enrichment was achieved."

But what is currently populated into the spreadsheet is:
> A term that may be associated with a process-related ontology term.

I am not sure what the best long term fix would be some options could be:
- write our own version of jsonref that includes all attributes when following refs
- when merging the schemas when creating the SchemaParser, if any fields are module refs (that aren't ontologies), also merge that sub-module with the standalone schema

**Comment by mshadbolt** [comment]
This is currently deployed in dev but not yet in staging
